# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/nsFocusManager.cpp
# Commit: 8b92fb4e3f41
# Full Hash: 8b92fb4e3f41d856629818bd2ecfc25bed23a643
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2023-12-23 09:52:03
# Regressor Bug: 779684
# File Overlap Count: 1
# Description:
#   Bug 779684 - part 3: Make `nsFrameIterator` non-refcountable r=emilio
#   
#   Nobody shares an instance of `nsFrameIterator`.  Therefore, we can make it
#   non-refcountable.  Additionally, `nsVisualIterator` is not required because
#   it overrides only 4 methods which are really simple.  So, once we merge it
# ==============================================================================

diff -r f357b2287591 -r 8b92fb4e3f41 dom/base/nsFocusManager.cpp
--- a/dom/base/nsFocusManager.cpp	Sat Dec 23 05:39:38 2023 +0000
+++ b/dom/base/nsFocusManager.cpp	Sat Dec 23 05:39:39 2023 +0000
@@ -3226,14 +3226,13 @@
         domSelection && domSelection->GetAncestorLimiter()
             ? domSelection->GetAncestorLimiter()->GetPrimaryFrame()
             : nullptr;
-    RefPtr<nsFrameIterator> frameIterator = nsFrameIterator::Create(
-        presContext, startFrame, nsFrameIterator::Type::Leaf,
-        false,  // aVisual
-        false,  // aLockInScrollView
-        true,   // aFollowOOFs
-        false,  // aSkipPopupChecks
-        limiter);
-    MOZ_ASSERT(frameIterator);
+    nsFrameIterator frameIterator(presContext, startFrame,
+                                  nsFrameIterator::Type::Leaf,
+                                  false,  // aVisual
+                                  false,  // aLockInScrollView
+                                  true,   // aFollowOOFs
+                                  false,  // aSkipPopupChecks
+                                  limiter);
 
     nsIFrame* newCaretFrame = nullptr;
     nsIContent* newCaretContent = start;
@@ -3242,8 +3241,8 @@
       // Continue getting the next frame until the primary content for the
       // frame we are on changes - we don't want to be stuck in the same
       // place
-      frameIterator->Next();
-      newCaretFrame = frameIterator->CurrentItem();
+      frameIterator.Next();
+      newCaretFrame = frameIterator.CurrentItem();
       if (!newCaretFrame) {
         break;
       }
@@ -4194,17 +4193,16 @@
       getNextFrame = false;
     }
 
-    RefPtr<nsFrameIterator> frameIterator;
+    Maybe<nsFrameIterator> frameIterator;
     if (frame) {
       // For tab navigation, pass false for aSkipPopupChecks so that we don't
       // iterate into or out of a popup. For document naviation pass true to
       // ignore these boundaries.
-      frameIterator = nsFrameIterator::Create(
-          presContext, frame, nsFrameIterator::Type::PreOrder,
-          false,                  // aVisual
-          false,                  // aLockInScrollView
-          true,                   // aFollowOOFs
-          aForDocumentNavigation  // aSkipPopupChecks
+      frameIterator.emplace(presContext, frame, nsFrameIterator::Type::PreOrder,
+                            false,                  // aVisual
+                            false,                  // aLockInScrollView
+                            true,                   // aFollowOOFs
+                            aForDocumentNavigation  // aSkipPopupChecks
       );
       MOZ_ASSERT(frameIterator);
 